src/core/Akka/Event/LogFilter.cs

```diff
@@ -0,0 +1,325 @@
+    public override LogFilterDecision ShouldKeepMessage(LogFilterType part, string content)
+    {
+        if (part == LogFilterType.Source)
+            return _sourceRegex.IsMatch(content) ? LogFilterDecision.Drop : LogFilterDecision.Keep;
+        return LogFilterDecision.NoDecision;
+    }
+}
+
+public sealed class ExactMatchLogSourceFilter : LogFilterBase
+{
+    private readonly string _source;
+    private readonly StringComparison _comparison;
+
+    public ExactMatchLogSourceFilter(string source, StringComparison comparison = StringComparison.OrdinalIgnoreCase)
+    {
+        _source = source;
+        _comparison = comparison;
+    }
+
+    public override LogFilterType FilterType => LogFilterType.Source;
+
+    public override LogFilterDecision ShouldKeepMessage(LogFilterType part, string content)
+    {
+        if (part == LogFilterType.Source)
+            return content == _source ? LogFilterDecision.Drop : LogFilterDecision.Keep;
+        return LogFilterDecision.NoDecision;
+    }
+}
+
+public sealed class RegexLogMessageFilter : LogFilterBase
+{
+    private readonly Regex _messageRegex;
+
+    public RegexLogMessageFilter(Regex messageRegex)
+    {
+        _messageRegex = messageRegex;
+    }
+
+    public override LogFilterType FilterType => LogFilterType.Message;
+
+    public override LogFilterDecision ShouldKeepMessage(LogFilterType part, string content)
+    {
+        if (part == LogFilterType.Message)
+            return _messageRegex.IsMatch(content) ? LogFilterDecision.Drop : LogFilterDecision.Keep;
+        return LogFilterDecision.NoDecision;
+    }
+}
+
+/// <summary>
+/// Runs inside the logging actor and evaluates if a log message should be kept.
+/// </summary>
+public class LogFilterEvaluator
+{
+    public static readonly LogFilterEvaluator NoFilters = EmptyLogFilterEvaluator.Instance;
+
+    private readonly LogFilterBase[] _filters;
+
+
+    /// <summary>
+    /// "Fast path" indicator - if this is true, we only evaluate log sources and not the message content.
+    /// </summary>
+    public bool EvaluatesLogSourcesOnly { get; }
+
+    public LogFilterEvaluator(LogFilterBase[] filters)
+    {
+        _filters = filters;
+        EvaluatesLogSourcesOnly = filters.All(x => x.FilterType == LogFilterType.Source);
+    }
+
+    public virtual bool ShouldTryKeepMessage(LogEvent evt, out string expandedLogMessage)
+    {
+        expandedLogMessage = string.Empty;
+
+        // fast and slow paths available here
+        if (EvaluatesLogSourcesOnly)
+        {
+            foreach (var filter in _filters)
+            {
+                // saves on allocations in negative cases, where we can avoid expanding the message
+                if (filter.ShouldKeepMessage(LogFilterType.Source, evt.LogSource) == LogFilterDecision.Drop)
+                    return false;
+            }
+        }
+        else
+        {
+            // allocate the message just once
+            var nullCheck = evt.Message.ToString();
```

Comment:
Scenario where we have to consider at least 1 rule pertaining to the body of the log message - unavoidable allocation here even in scenarios where we drop the log.